app/controllers/collections_controller.rb

```diff
@@ -338,4 +351,80 @@ def write_item_metadata(corpus_dir, item_json)
+    item_subject = RDF::URI.new(item.uri)
+    item_query = RDF::Query.new do
+      pattern [item_subject, :predicate, :object]
+    end
+    item_statements = repository.query(item_query)
+    item_statements.each do |item_statement|
+      repository.delete(RDF::Statement(item_subject, item_statement[:predicate], item_statement[:object]))
+    end
+  end
+
+  # Deletes statements with the document's derived URI from Sesame
+  def delete_document_from_sesame(document, repository)
+    document_query = RDF::Query.new do
+      pattern [:subject, MetadataHelper::SOURCE, RDF::URI.new("file://#{document.file_path}")]
+      pattern [:subject, MetadataHelper::IDENTIFIER, "#{document.file_name}"]
+    end
+    document_URIs = repository.query(document_query)
+    if document_URIs.count == 1
+      document_URI = document_URIs.first[:subject]
+      document_statements_query = RDF::Query.new do
+        pattern [document_URI, :predicate, :object]
+      end
+      document_statements = repository.query(document_statements_query)
+      document_statements.each do |document_statement|
+        repository.delete(RDF::Statement(document_URI, document_statement[:predicate], document_statement[:object]))
+      end
+    else
+      Rails.logger.error "Cannot delete document RDF as multiple distinct document URIs match the document file name and path"
+      document_URIs.each do |statement|
+        Rails.logger.debug "#{RDF::Statement(statement[:subject], MetadataHelper::SOURCE, RDF::URI.new("file://#{document.file_path}"))}"
+        Rails.logger.debug "#{RDF::Statement(statement[:subject], MetadataHelper::IDENTIFIER, "#{document.file_name}")}"
+      end
+    end
+  end
+
+  # Removes an item and its documents from the database, filesystem, Sesame and Solr
+  def remove_item(item, collection, corpus_dir)
+    # Remove from filesystem
+    item_name = item.handle.split(":")[1]
+    delete_file(File.join(corpus_dir, "#{item_name}-metadata.rdf"))
+    item.documents.each do |document|
+      delete_file(document.file_path)
+    end
+    # Remove from Sesame
+    server = RDF::Sesame::HcsvlabServer.new(SESAME_CONFIG["url"].to_s)
+    repository = server.repository(collection.name)
+    delete_item_from_sesame(item, repository)
+    item.documents.each do |document|
+      delete_document_from_sesame(document, repository)
+    end
+    # Remove from Solr
+    stomp_client = Stomp::Client.open "stomp://localhost:61613"
+    deindex_item_from_solr(item.id, stomp_client)
+    stomp_client.close
+    # Remove from database (item, its documents and their document audits)
+    item.destroy
+  end
+
+  # Attempts to delete a file or logs any exceptions raised
+  def delete_file(file_path)
+    begin
+      if File.exists?(file_path)
+        File.delete(file_path)
+      else
+        Rails.logger.error "Could not find file to delete: #{file_path}"
+        false
+      end
+    rescue => e
+      Rails.logger.error e.inspect
```

Comment:
Be careful about trapping errors here and not re-raising them. Your main controller method traps errors and response with an error message. If Solr or Sesame run into any issues, the user will know about it, but not in the case of filesystem errors (e.g. if there were permission errors for some reason and the file couldn't be deleted.)